core/src/log/elastic.rs

```diff
@@ -0,0 +1,188 @@
+    pub general: ElasticLoggingConfig,
+    /// The Elasticsearch index to log to.
+    pub index: String,
+    /// The Elasticsearch type to use for logs.
+    pub document_type: String,
+    /// The subgraph ID that the drain is for.
+    pub subgraph_id: String,
+}
+
+/// An slog `Drain` for logging to Elasticsearch.
+pub struct ElasticDrain {
+    config: ElasticDrainConfig,
+}
+
+impl ElasticDrain {
+    pub fn new(config: ElasticDrainConfig) -> Self {
+        ElasticDrain { config }
+    }
+}
+
+impl Drain for ElasticDrain {
+    type Ok = ();
+    type Err = reqwest::Error;
+
+    fn log(&self, record: &Record, values: &OwnedKVList) -> StdResult<Self::Ok, Self::Err> {
+        let timestamp = Utc::now().to_rfc3339_opts(SecondsFormat::Nanos, true);
+        let id = format!("{}-{}", self.config.subgraph_id, timestamp);
+
+        // Serialize logger arguments
+        let mut serializer = SimpleKVSerializer::new();
+        record
+            .kv()
+            .serialize(record, &mut serializer)
+            .expect("failed to serializer logger arguments");
+        let (n_logger_kvs, logger_kvs) = serializer.finish();
+
+        // Serialize log message arguments
+        let mut serializer = SimpleKVSerializer::new();
+        values
+            .serialize(record, &mut serializer)
+            .expect("failed to serialize log message arguments");
+        let (n_value_kvs, value_kvs) = serializer.finish();
+
+        let mut text = format!("{}", record.msg());
+        if n_logger_kvs > 0 {
+            write!(text, ", {}", logger_kvs);
+        }
+        if n_value_kvs > 0 {
+            write!(text, ", {}", value_kvs);
+        }
+
+        let meta = ElasticLogMeta {
+            module: record.module().into(),
+            line: record.line() as i64,
+            column: record.column() as i64,
+        };
+
+        let log = ElasticLog {
+            id: id.clone(),
+            subgraph_id: self.config.subgraph_id.clone(),
+            timestamp,
+            text,
+            level: record.level(),
+            meta,
+        };
+
+        // Build the document URL
+        let mut document_url = reqwest::Url::parse(self.config.general.endpoint.as_str())
+            .expect("invalid Elasticsearch URL");
+        document_url
+            .path_segments_mut()
+            .expect("failed to set the Elasticsearch documenet path")
+            .push(self.config.index.as_str())
+            .push(self.config.document_type.as_str())
+            .push(id.as_str());
+
+        // Send log to Elasticsearch
+        let client = reqwest::Client::new();
+        let response = client
+            .put(document_url)
+            .basic_auth(
+                self.config.general.username.clone(),
+                self.config.general.password.clone(),
+            ).json(&log)
+            .send()?;
+
+        // Return an error if the server returned an error response
+        response.error_for_status().map(|_| ())
```

Comment:
Is the response completely worthless here? Not worth logging (to a logger that doesn't go to ES, of course)?